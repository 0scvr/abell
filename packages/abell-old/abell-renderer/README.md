# Abell Renderer

A template engine used in [Abell](https://github.com/abelljs/abell).

**Documentation** - [https://abelljs.org/abell-renderer/](https://abelljs.org/abell-renderer/)