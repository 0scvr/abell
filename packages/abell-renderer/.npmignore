src/
tsconfig.tsbuildinfo
tsconfig.json
ConceptNote.md