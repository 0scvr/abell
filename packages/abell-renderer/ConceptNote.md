## Single File Components

```jsx
<AbellComponent name="Description">
  <p>{{ props.children }}</p>
</AbellComponent>

<html>
<body>
  <h2>Hello, World</h2>
  <Description>My description text</Description>
</body>
</html>
```

## Compile to JS?

```jsx
{{
  function Description(props) {
    return `
      <p>${props.children}</p>
    `
  }
}}

<html>
<body>
  <h2>Hello, World</h2>
  {{ Description({children: 'My description text'}) }}
</body>
</html>
```
