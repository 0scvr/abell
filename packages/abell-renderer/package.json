{
  "name": "abell-renderer",
  "version": "1.0.0-alpha.2",
  "description": "",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "ts-node ./src/example.ts",
    "dev": "yarn example:dev",
    "example:dev": "yarn ae dev",
    "dev:with-ts-run": "ts-run src/example.ts -w",
    "dev:with-esbuild": "yarn esbuild",
    "ts-node": "ts-node",
    "esbuild": "node scripts/dev-build.js && node dev/example/index.js",
    "build": "tsc",
    "serve": "yarn build && node dist/example.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "acorn": "^8.3.0",
    "htmlparser2": "^6.1.0",
    "postcss-selector-parser": "^6.0.6",
    "stylis": "^4.0.10"
  },
  "devDependencies": {
    "@types/stylis": "^4.0.1"
  }
}
