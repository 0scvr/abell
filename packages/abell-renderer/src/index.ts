import fs from 'fs';
import path from 'path';

import { UserOptions, AbellComponentMap } from './types';
import { getAbellInBuiltSandbox } from './utils';
import abellParser from './parse-abell';

const defaultUserOptions: UserOptions = {
  basePath: '',
  allowRequire: false,
  allowComponents: false,
  filename: '<sandbox>.abell'
};

export function render(
  abellTemplate: string,
  // eslint-disable-next-line @typescript-eslint/ban-types
  userSandbox: object = {},
  options: UserOptions = defaultUserOptions
): AbellComponentMap {
  // Set initial variables
  options.basePath =
    options.basePath ||
    (options.filename && path.dirname(options.filename)) ||
    '';

  const builtInFunctions = getAbellInBuiltSandbox(options);

  // console.log(builtInFunctions);
  abellParser.write(abellTemplate);
  abellParser.end();

  // Add built-in functions along with user's sandbox

  return {
    html: ''
  };
}

export default {
  render
};
