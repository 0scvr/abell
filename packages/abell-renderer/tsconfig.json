{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    /* Basic Options */
    "outDir": "dist",
    "rootDir": "src",
    "composite": true
  },
  "include": ["src"],
  "exclude": ["dist"]
}