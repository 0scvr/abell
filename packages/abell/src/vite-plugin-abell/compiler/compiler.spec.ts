/* eslint-disable max-len */
import path from 'path';
import { describe, test, expect } from 'vitest';
import { compile } from './index';

describe('compile()', () => {
  test('should successfully compile single expressions', () => {
    const out = compile('<body>{{ 3 + 4 }}</body>', {
      filepath: __dirname,
      outputType: 'syntax-blocks'
    });
    expect(out.out.text).toMatchInlineSnapshot('"<body>${e( 3 + 4 )}</body>"');
    expect(out.declarationsBlock.text).toMatchInlineSnapshot('""');
  });

  test('should create first block as declaration block with declarations comment', () => {
    const abellCode = `
    {{
      // declarations
      const a = 3;
    }}
    <body>{{ a }}</body>
    `;
    const out = compile(abellCode, {
      filepath: __dirname,
      outputType: 'syntax-blocks'
    });
    expect(out.out.text.trim()).toMatchInlineSnapshot(
      '"<body>${e( a )}</body>"'
    );
    expect(out.declarationsBlock.text.trim()).toMatchInlineSnapshot(`
      "// declarations
            const a = 3;"
    `);
  });

  test('should successfully compile with declarations', () => {
    const abellCode = `
    {{
      import fs from 'fs';
      import path from 'path';
      import { compile } from 'abell-renderer';
    }}
    <body>
      {{ 3 + 4 }}
      <b>{{ 'Helloo'.toUpperCase() }}</b>
    </body>
    `;
    const out = compile(abellCode, {
      filepath: __dirname,
      outputType: 'syntax-blocks'
    });
    expect(out.declarationsBlock.text).toMatchInlineSnapshot(`
      "
            import fs from 'fs';
            import path from 'path';
            import { compile } from 'abell-renderer';
          "
    `);
    expect(out.out.text.trim()).toMatchInlineSnapshot(`
      "<body>
            \${e( 3 + 4 )}
            <b>\${e( 'Helloo'.toUpperCase() )}</b>
          </body>"
    `);
  });
});

describe('scoped css', () => {
  test('should scope the css', () => {
    const abellCode = `
    <nav>hello</nav>

    <style export>
    nav {
      color: yellow;
    }
    </style>

    <style hello="hi">
    nav {
      color: blue;
    }
    </style>

    <style scoped="false">
    nav {
      color: red;
    }
    </style>
    `;

    const { out } = compile(abellCode, {
      filepath: path.join(process.cwd(), 'test.abell'),
      outputType: 'syntax-blocks'
    });
    expect(out.text.trim()).toMatchInlineSnapshot(`
      "<style abell-generated>nav[data-abell-duqnjx]{color:yellow;}</style><style abell-generated>nav[data-abell-duqnjx]{color:blue;}</style><style abell-ignored>
          nav {
            color: red;
          }
          </style>
          <nav data-abell-duqnjx>hello</nav>"
    `);
  });

  test('should not scope the css on html page', () => {
    const abellCode = `
    <html>
    <head>
    <style>
    nav {
      color: yellow;
    }
    </style>
    </head>
    <body>
      <nav>hello</nav>
    </body>
    </html>
    `;

    const { out } = compile(abellCode, {
      filepath: path.join(process.cwd(), 'test.abell'),
      outputType: 'syntax-blocks'
    });
    expect(out.text.trim()).toMatchInlineSnapshot(`
      "<html>
          <head>
          <style>
          nav {
            color: yellow;
          }
          </style>
          </head>
          <body>
            <nav>hello</nav>
          </body>
          </html>"
    `);
  });

  test('should not scope the css when there are no style tags', () => {
    const abellCode = `
    <nav>hello</nav>
    `;

    const { out } = compile(abellCode, {
      filepath: path.join(process.cwd(), 'test.abell'),
      outputType: 'syntax-blocks'
    });
    expect(out.text.trim()).toMatchInlineSnapshot('"<nav>hello</nav>"');
  });

  test('should not scope the css when there are no scoped style tags', () => {
    const abellCode = `
    <nav>hello</nav>
    <style scoped="false">
    nav {
      color: red;
    }
    </style>
    `;

    const { out } = compile(abellCode, {
      filepath: path.join(process.cwd(), 'test.abell'),
      outputType: 'syntax-blocks'
    });

    expect(out.text.trim()).toMatchInlineSnapshot(`
      "<style abell-ignored>
          nav {
            color: red;
          }
          </style>
          <nav>hello</nav>"
    `);
  });

  test('should handle abell blocks inside style tag', () => {
    const abellCode = `
    {{
      import x from '.xyz'
    }}
    <html>
    <head>
    <style>
    {{ 2 + 1 }}
    </style>
    </head>
    <body>
      <nav>hello</nav>
    </body>
    </html>
    `;

    const { out } = compile(abellCode, {
      filepath: path.join(process.cwd(), 'test.abell'),
      outputType: 'syntax-blocks'
    });
    expect(out.text.trim()).toMatchInlineSnapshot(`
      "<html>
          <head>
          <style>
          \${e( 2 + 1 )}
          </style>
          </head>
          <body>
            <nav>hello</nav>
          </body>
          </html>"
    `);
  });
});
