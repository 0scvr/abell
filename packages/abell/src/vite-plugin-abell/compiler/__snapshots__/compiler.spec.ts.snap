// Vitest Snapshot v1

exports[`compile() > should successfully compile single expressions 1`] = `
"import { default as _path } from 'path';

const __filename = '/Users/saurabhdaware/Desktop/projects/abell-org/abell/packages/abell/src/vite-plugin-abell/compiler';
const __dirname = _path.dirname(__filename);
const e = (val) => {
  if (typeof val === 'function') return val();
  if (!val) return '';
  return val;
};
export const html = (props = {}) => {
  const Abell = { 
    props, 
    console: {
      log: (...args) => {
        console.log('\\\\u001b[90m[src/vite-plugin-abell/compiler]:\\\\u001b[39m', ...args);
        return '';
      }
    }
  };
  return \`<body>\${e( 3 + 4 )}</body>\`
};
export default html;"
`;

exports[`compile() > should successfully compile with declarations 1`] = `
"import { default as _path } from 'path';

      import fs from 'fs';
      import path from 'path';
      import { compile } from 'abell-renderer';
    
const __filename = '/Users/saurabhdaware/Desktop/projects/abell-org/abell/packages/abell/src/vite-plugin-abell/compiler';
const __dirname = _path.dirname(__filename);
const e = (val) => {
  if (typeof val === 'function') return val();
  if (!val) return '';
  return val;
};
export const html = (props = {}) => {
  const Abell = { 
    props, 
    console: {
      log: (...args) => {
        console.log('\\\\u001b[90m[src/vite-plugin-abell/compiler]:\\\\u001b[39m', ...args);
        return '';
      }
    }
  };
  return \`<body>
      \${e( 3 + 4 )}
      <b>\${e( 'Helloo'.toUpperCase() )}</b>
      <div>
        \${e(
          [1, 2, 3].map(i => 
            i * 2
          )
        )}
      </div>
    </body>\`
};
export default html;"
`;
